src/controller/python/chip/ChipDeviceCtrl.py

```diff
@@ -1427,3 +1269,203 @@ def _InitLib(self):
+                c_void_p), c_uint64, c_uint64, c_uint16, c_char_p, c_bool, c_bool, POINTER(c_uint32), c_uint32]
+            self._dmLib.pychip_OpCreds_AllocateController.restype = PyChipError
+
+            self._dmLib.pychip_OpCreds_AllocateControllerForCustomCommissioningFlow.argtypes = [
+                POINTER(c_void_p), c_char_p, c_uint32, c_char_p, c_uint32, c_char_p, c_uint32, c_char_p, c_uint32, c_uint16, c_bool]
+            self._dmLib.pychip_OpCreds_AllocateControllerForCustomCommissioningFlow.restype = PyChipError
+
+
+class ChipDeviceController(ChipDeviceControllerBase):
+    def __init__(self, opCredsContext: ctypes.c_void_p, fabricId: int, nodeId: int, adminVendorId: int, catTags: typing.List[int] = [], paaTrustStorePath: str = "", useTestCommissioner: bool = False, fabricAdmin: FabricAdmin = None, name: str = None):
+        super().__init__("caIndex(%x)/fabricId(0x%016X)/nodeId(0x%016X)" % (fabricAdmin.caIndex, fabricId, nodeId) if name is None else name)
+
+        self._dmLib.pychip_DeviceController_SetIssueNOCChainCallbackPythonCallback(_IssueNOCChainCallbackPythonCallback)
+
+        devCtrl = c_void_p(None)
+
+        c_catTags = (c_uint32 * len(catTags))()
+
+        for i, item in enumerate(catTags):
+            c_catTags[i] = item
+
+        self._dmLib.pychip_OpCreds_AllocateController.argtypes = [c_void_p, POINTER(
+            c_void_p), c_uint64, c_uint64, c_uint16, c_char_p, c_bool, c_bool, POINTER(c_uint32), c_uint32]
+        self._dmLib.pychip_OpCreds_AllocateController.restype = PyChipError
+
+        # TODO(erjiaqing@): Figure out how to control enableServerInteractions for a single device controller (node)
+        self._ChipStack.Call(
+            lambda: self._dmLib.pychip_OpCreds_AllocateController(c_void_p(
+                opCredsContext), pointer(devCtrl), fabricId, nodeId, adminVendorId, c_char_p(None if len(paaTrustStorePath) == 0 else str.encode(paaTrustStorePath)), useTestCommissioner, self._ChipStack.enableServerInteractions, c_catTags, len(catTags))
+        ).raise_on_error()
+
+        self._fabricAdmin = fabricAdmin
+        self._fabricId = fabricId
+        self._nodeId = nodeId
+        self._caIndex = fabricAdmin.caIndex
+
+        self._set_dev_ctrl(devCtrl=devCtrl)
+
+        self._finish_init()
+
+        assert self._fabricId == fabricId
+        assert self._nodeId == nodeId
+
+    @property
+    def caIndex(self) -> int:
+        return self._caIndex
+
+    @property
+    def fabricAdmin(self) -> FabricAdmin:
+        return self._fabricAdmin
+
+    def Commission(self, nodeid):
+        self.CheckIsActive()
+        self._ChipStack.commissioningCompleteEvent.clear()
+        self.state = DCState.COMMISSIONING
+
+        self._ChipStack.CallAsync(
+            lambda: self._dmLib.pychip_DeviceController_Commission(
+                self.devCtrl, nodeid)
+        )
+        if not self._ChipStack.commissioningCompleteEvent.isSet():
+            # Error 50 is a timeout
```

Comment:
what does "error 50" mean in this comment? I read the code as `if commissioning complete event  is not set, return false` ... nowhere anything about a error 50 or a timeout.

Reply:
done